### PR TITLE
feat(snapshot, stages): snapshot stage

### PR DESCRIPTION
**Pull request by shekhirin**
There's a scenario when we have data in the database for `finalized..=tip` blocks, turn off the node, turn it on again, and pipeline starts, writing to static files. The problem is that data in the database was never moved to static files.

This PR introduces a separate pipeline stage that copies all (even non-finalized, and it's fine because we can unwind from static files) data from the database to static files. The actual deletion will be done by pruner, as before.

https://github.com/paradigmxyz/reth/blob/fb77c957bf20e013bbb1645b653a149567e62275/crates/stages/src/stages/snapshot.rs#L7-L10

---

Tested on Holesky by taking a database generated with `main` branch, unwinding it by 1000 blocks, and starting the node on `alexey/snapshot-stage` branch. It triggers the pipeline, and the `Snapshot` stage moves all data from database to static files.

### CODE REVIEW

File: crates/stages/src/stages/snapshot.rs [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+        self.snapshotter.run(targets)?;
+        Ok(ExecOutput::done(input.checkpoint()))
+    }
+
+    fn unwind(
+        &mut self,
+        _provider: &DatabaseProviderRW<DB>,
+        _input: UnwindInput,
+    ) -> Result<UnwindOutput, StageError> {
+        Err(StageError::UnwindNotSupported)
```
Comment (mattsse) [inline crates/stages/src/stages/snapshot.rs:43]: doesn't this result in unwind errors?

**Review by mattsse — APPROVED** [review]
this seems reasonable, pending @onbjerg

File: crates/stages/src/stages/snapshot.rs (hunk 1 above)
Comment (onbjerg) [reply crates/stages/src/stages/snapshot.rs:43]: same question - will this not block unwinding from working altogether? if unwind is supposed to be a noop here i'd rather we return an Ok(..) as if we did the unwind

**Review by onbjerg — APPROVED** [review]
lgtm minus the comment @mattsse left

File: crates/stages/src/stages/snapshot.rs (hunk 1 above)
Comment (shekhirin) [reply crates/stages/src/stages/snapshot.rs]: that's true, need an `Ok(..)` here. fixed

**Pull request merged**